api/pkg/auth/service/auth.go

```diff
@@ -0,0 +1,201 @@
+func Authenticate(res http.ResponseWriter, req *http.Request) {
+	UI_URL = req.FormValue("redirect_uri")
+	gothic.BeginAuthHandler(res, req)
+}
+
+// Once user is authenticated, store the user details in db
+// and redirect to UI with the status code and auth code
+func (s *service) AuthCallBack(res http.ResponseWriter, req *http.Request) {
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	ghUser, err := gothic.CompleteUserAuth(res, req)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	params := req.URL.Query()
+
+	if err = r.insertData(ghUser, params.Get("code")); err != nil {
+		r.log.Error(err)
+		res.Header().Set("Location", fmt.Sprintf("%s?status=%d", UI_URL, http.StatusBadRequest))
+		res.WriteHeader(http.StatusTemporaryRedirect)
+	}
+
+	res.Header().Set("Location", fmt.Sprintf("%s?status=%d&code=%s", UI_URL, http.StatusOK, params.Get("code")))
+	res.WriteHeader(http.StatusTemporaryRedirect)
+}
+
+// Checks for auth code in the headers, validates the
+// auth code and returns the jwt token for user
+func (s *service) HubAuthenticate(res http.ResponseWriter, req *http.Request) {
+
+	// Get the auth code from params
+	code := req.FormValue("code")
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	var user model.User
+	// Check if user exist
+	q := r.db.Model(&model.User{}).
+		Where("code = ?", code)
+
+	err := q.First(&user).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
+			r.log.Error(err)
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		} else {
+			r.log.Error(err)
+			http.Error(res, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	if err := r.db.Model(&model.User{}).Where("github_login = ?", user.GithubLogin).Update("code", "").Error; err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// gets user scopes to add in jwt
+	scopes, err := r.userScopes(&user)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	userTokens, err := r.createTokens(&user, scopes)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	res.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(res).Encode(userTokens); err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+// Provides a list of git provider present in auth server
+func List(res http.ResponseWriter, req *http.Request) {
+	providers := authApp.ProviderList{
+		Data: []authApp.Provider{
+			{
```

Comment:
can we add a `TODO` comment over here to configure these values dynamically?

Reply:
when we want to add a new provide let say google? will there be code change?

Reply:
yes

Reply:
nope

Reply:
I mean the logic for adding the provider for auth